fineract-provider/src/main/java/org/apache/fineract/cob/service/InlineLoanCOBExecutorServiceImpl.java

```diff
@@ -0,0 +1,137 @@
+    }
+
+    public List<Long> getLoanIds() {
+        return loanIds;
+    }
+
+    private List<LoanAccountLock> getLoanAccountLocks(List<Long> loanIds) {
+        List<LoanAccountLock> loanAccountLocks = new ArrayList<>();
+        List<Long> alreadyLockedLoanIds = new ArrayList<>();
+        loanIds.forEach(loanId -> {
+            Optional<LoanAccountLock> loanLockOptional = loanAccountLockRepository.findById(loanId);
+            if (loanLockOptional.isPresent()) {
+                LoanAccountLock loanAccountLock = loanLockOptional.get();
+                if (isLockOverruleable(loanAccountLock)) {
+                    loanAccountLocks.add(loanAccountLock);
+                } else {
+                    alreadyLockedLoanIds.add(loanId);
+                }
+            } else {
+                loanAccountLocks.add(new LoanAccountLock(loanId, LockOwner.LOAN_INLINE_COB_PROCESSING));
+            }
+        });
+        if (!alreadyLockedLoanIds.isEmpty()) {
+            String message = "There is a hard lock on the loan account without any error, so it can't be overruled.";
+            String loanIdsMessage = " Locked loan IDs: " + alreadyLockedLoanIds;
+            throw new LoanAccountLockCannotBeOverruledException(message + loanIdsMessage);
+        }
+        return loanAccountLocks;
+    }
+
+    private void execute(List<LoanAccountLock> loanAccountLocks, List<Long> loanIds, String jobName) {
+        lockLoanAccounts(loanAccountLocks);
+        this.loanIds = loanIds;
+        Job inlineLoanCOBJob;
+        try {
+            inlineLoanCOBJob = jobLocator.getJob(jobName);
+        } catch (NoSuchJobException e) {
+            throw new JobNotFoundException(jobName, e);
+        }
+        JobParameters jobParameters = new JobParametersBuilder(jobExplorer).getNextJobParameters(inlineLoanCOBJob)
+                .addJobParameters(new JobParameters()).toJobParameters();
+        JobExecution jobExecution;
+        try {
+            jobExecution = jobLauncher.run(inlineLoanCOBJob, jobParameters);
+        } catch (Exception e) {
+            final String msg = "Job execution failed for job with name:" + jobName;
+            log.error("{}", msg, e);
+            throw new PlatformInternalServerException("error.msg.sheduler.job.execution.failed", msg, jobName, e);
+        }
+        if (!BatchStatus.COMPLETED.equals(jobExecution.getStatus())) {
+            final String msg = "Job execution failed for job with name:" + jobName;
+            log.error("{}", msg);
+            throw new PlatformInternalServerException("error.msg.sheduler.job.execution.failed", msg, jobName);
+        }
+    }
+
+    private void lockLoanAccounts(List<LoanAccountLock> loanAccountLocks) {
+        loanAccountLocks.forEach(loanAccountLock -> {
+            loanAccountLock.setNewLockOwner(LockOwner.LOAN_INLINE_COB_PROCESSING);
+            loanAccountLockRepository.save(loanAccountLock);
```

Comment:
So the public method is not allowing transactions which means that this save call is opening and closing transactions for every single loan account lock, that's definitely not the way to go. All loan accounts shall be locked in a single tx